posts/statistical-correctness/index.qmd

```diff
@@ -0,0 +1,146 @@
+
+[^1]: But see the ["testing statistical software" post from Alex Hayes](https://www.alexpghayes.com/post/2019-06-07_testing-statistical-software/) where he presents his process to determine if he deems a statistical package trustworthy or not.
+
+In this blog post, we explore various methods to ensure the correctness of your statistical package. Importantly, we argue that these tests should be part of your continuous integration system, to ensure your tools remains valid throughout its life, and to let users verify how you validate your package.
+
+The approaches presented here are non-exclusive and should ideally all be added to your tests. However, they are presented in order of stringency and priority to implement. We also take a example of a function computing the centroid of a list of points to demonstrate how you would follow the recommendations presentation here:
+
+```{r}
+#' Compute the centroid of a set of points
+#'
+#' @coords Coordinates of the points as a list of vectors. Each element of the 
+#'   list is a point.
+#'
+#' @returns A vector of coordinates of the same length of each element of 
+#'   `coords`
+#'   
+#' @examples
+#' centroid(list(c(0, 1, 5, 3), c(8, 6, 4, 3), c(10, 2, 3, 7)))
+#' 
+centroid <- function(coords) {
+
+  # ...
+  # Skip all the necessary input checking for the purpose of this demo
+  # ...
+
+  coords_mat <- do.call(rbind, coords)
+  
+  return(colMeans(coords_mat))
+  
+}
+```
+
+# Compare your results to the reference implementation
+
+The most straightforward and most solid way to ensure your implementation is valid is to compare your results to the results of the reference implementation. The reference implementation can be a package in another language, an example with toy data in the scientific article introducing the method, etc.
+
+For example, the [`{gemma2}` R package](https://github.com/fboehm/gemma2), which re-implements the methods from [the GEMMA tool written in C++](https://github.com/genetics-statistics/GEMMA), [verifies that values produced by both tools match](https://github.com/fboehm/gemma2/blob/ea3052f8609622f17224fb8ec5fd83bd1bceb33e/tests/testthat/test_calc_sigma.R#L34-L37):
+
+``` r
+test_that("Results of gemma2 equal those of GEMMA v 0.97", {
+  expect_equal(Sigma_ee, diag(c(18.559, 12.3672)), tolerance = 0.0001)
+  expect_equal(Sigma_uu, diag(c(82.2973, 41.9238)), tolerance = 0.0001)
+})
+```
+
+Note that even if a **reference** implementation doesn't exist, it is still good practice to compare your implementation to competing ones. Discrepancies might reveal a bug in your implementation or theirs but in any case, finding it out is beneficial to the community.
+
+However, this approach cannot be used in all cases. Indeed, there may not be a reference implementation in your case. Or it might be difficult to replicate identical computations in the case of algorithm with stochasticity [^2].
```

Comment:
The `kind` argument in `set.seed()` might help in this situation but I agree that RNG across languages will make replication difficult.

Reply:
Even then, I'm not sure it's possible to generate the same random number. In addition with the kind of RNG, you have the initialization problem, and `?set.seed()` states that they have custom initialization methods in some cases. After this, how the `.Random.seed` is used in functions like `random.random()` in python or `runif()` in R may differ.

I would be very interested in hearing if you manage to create the same random value in two languages though :+1: